src/webui/api/searchcontroller.cpp

```diff
@@ -0,0 +1,384 @@
+
+    auto searchHandlers = session->getData(QLatin1String("searchHandlers")).value<SearchHandlerDict>();
+    searchHandlers.insert(id, searchHandler);
+    session->setData(QLatin1String("searchHandlers"), QVariant::fromValue(searchHandlers));
+
+    activeSearches.insert(id);
+    session->setData(QLatin1String("activeSearches"), QVariant::fromValue(activeSearches));
+
+    const int hour = 3600000;
+    QTimer::singleShot(hour, this, [id, this] () {searchTimeout(id);});
+
+    const QJsonObject result = {{"id", id}};
+    setResult(result);
+}
+
+void SearchController::stopAction()
+{
+    checkParams({"id"});
+
+    const QString id = params()["id"].trimmed();
+    ISession *session = sessionManager()->session();
+
+    auto searchHandlers = session->getData(QLatin1String("searchHandlers")).value<SearchHandlerDict>();
+    if (!searchHandlers.contains(id))
+        throw APIError(APIErrorType::NotFound);
+
+    SearchHandlerPtr searchHandler = searchHandlers[id];
+
+    if (searchHandler->isActive())
+    {
+        searchHandler->cancelSearch();
+        auto activeSearches = session->getData(QLatin1String("activeSearches")).value<QSet<QString>>();
+        activeSearches.remove(id);
+        session->setData(QLatin1String("activeSearches"), QVariant::fromValue(activeSearches));
+    }
+}
+
+void SearchController::statusAction()
+{
+    checkParams({"id"});
+
+    const QString id = params()["id"].trimmed();
+
+    auto searchHandlers = sessionManager()->session()->getData(QLatin1String("searchHandlers")).value<SearchHandlerDict>();
+    if (!searchHandlers.contains(id))
+        throw APIError(APIErrorType::NotFound);
+
+    SearchHandlerPtr searchHandler = searchHandlers[id];
+
+    const QJsonObject result = {
+        {"status", searchHandler->isActive() ? "Running" : "Stopped"},
+        {"total", searchHandler->results().size()}
+    };
+    setResult(result);
+}
+
+void SearchController::resultsAction()
+{
+    checkParams({"id"});
+
+    const QString id = params()["id"].trimmed();
+    int limit = params()["limit"].toInt();
+    int offset = params()["offset"].toInt();
+
+    auto searchHandlers = sessionManager()->session()->getData(QLatin1String("searchHandlers")).value<SearchHandlerDict>();
+    if (!searchHandlers.contains(id))
+        throw APIError(APIErrorType::NotFound);
+
+    SearchHandlerPtr searchHandler = searchHandlers[id];
+    const QList<SearchResult> searchResults = searchHandler->results();
+    const int size = searchResults.size();
+
+    if ((offset > size) || ((offset == size) && (offset != 0)) || (limit < 0))
+        throw APIError(APIErrorType::Conflict, tr("Offset or limit is out of range"));
+
+    // normalize values
+    if (offset < 0)
+        offset = size + offset;
+    if (offset < 0)  // check again
+        throw APIError(APIErrorType::Conflict);
+    if (limit == 0)
```

Comment:
What do you mean exactly?
And what do you suggest to change in line 138: `int limit = params()["limit"].toInt();`

The idea is treat `limit <= 0` as unlimited. But if we pass 0 to `QList::mid()` we will get empty result, since **QList** use 0 as is.

Reply:
I suggest to use only negative number as absence of the limit: `int limit = params().value("limit", "-1").toInt()`.

> …we will get empty result…

That's right. I would expect exactly that. As a side note, "limit" seems to me a bad choice for the parameters name, which is "count". "Limit" could mean "offset" + "count", the last index, and maybe something else.

Reply:
>As a side note, "limit" seems to me a bad choice for the parameter name, which is "count". "Limit" could mean "offset" + "count", the last index, and maybe something else.

You've probably never dealt with SQL? Limit means "upper bound on the number of rows returned by the entire SELECT statement" there. But you're certainly going to say something like "Stupid SQL...".
IMO, the "limit" term is better than "count", since it means "at most N values" but not "exactly N values" and that's exactly what we want here.
>That's right. I would expect exactly that.

Or let the user request a deliberately empty result subset is a good idea for you? Then explain why you need it.

Reply:
>Have no idea what are you referring to

I only referred to it to show this term might mean not only `"offset" + "count", the last index, and maybe something else`...
>can easily imagine scripting using the API and then 0 with a special meaning would add needless tests.

OK.

Reply:
> I only referred to it to show this term might mean not only 

That's exactly my point: "limit" might mean various things. Probably not a big deal for SQL, where one works with a particular dialect for years, but might be not so good for yet another application with its own API.

Reply:
> Don't need it right away, of course, but can easily imagine scripting using the API and then 0 with a special meaning would add needless tests.
 
That's why I left the `limit` parameter as optional. You only need to send it if you explicitly want to impose a limit.